File: initial_load.py
"""Utility to export historical Google Ads click data from Snowflake."""

import os
import json
import boto3
import snowflake.connector

# Environment variables for Snowflake
SF_ACCOUNT = os.environ['SNOWFLAKE_ACCOUNT']
SF_USER = os.environ['SNOWFLAKE_USER']
SF_PASSWORD = os.environ['SNOWFLAKE_PASSWORD']
SF_WAREHOUSE = os.environ['SNOWFLAKE_WAREHOUSE']
SF_DATABASE = os.environ['SNOWFLAKE_DATABASE']
SF_SCHEMA = os.environ['SNOWFLAKE_SCHEMA']

S3_BUCKET = os.environ['S3_BUCKET']
S3_PREFIX = os.environ.get('S3_KEY_PREFIX', 'click_performance/')
DDB_TABLE = os.environ['DYNAMO_TABLE_NAME']

TABLE_NAME = 'SEGMENT_EVENTS.GOOGLE_ADS_CLICK_PARAMETERS.CLICK_PERFORMANCE_REPORTS'

s3 = boto3.client('s3')
ddb = boto3.resource('dynamodb').Table(DDB_TABLE)


def fetch_all_rows():
    """Generator yielding all rows from the Snowflake table as dicts."""
    ctx = snowflake.connector.connect(
        user=SF_USER,
        password=SF_PASSWORD,
        account=SF_ACCOUNT,
        warehouse=SF_WAREHOUSE,
        database=SF_DATABASE,
        schema=SF_SCHEMA,
    )
    cs = ctx.cursor()
    try:
        query = f"SELECT * FROM {TABLE_NAME}"
        cs.execute(query)
        for row in cs.fetchall():
            yield dict(zip([c[0] for c in cs.description], row))
    finally:
        cs.close()
        ctx.close()


def write_to_dynamodb(rows):
    """Write the given rows to DynamoDB in batches."""
    with ddb.batch_writer() as batch:
        for r in rows:
            batch.put_item(Item=r)


def dump_to_s3(rows, key):
    """Upload the rows to S3 as a JSON document."""
    s3.put_object(Bucket=S3_BUCKET, Key=key, Body=json.dumps(rows))


if __name__ == '__main__':
    """Export the entire table and load it into S3 and DynamoDB."""
    rows = list(fetch_all_rows())
    key = f"{S3_PREFIX}initial_load.json"
    dump_to_s3(rows, key)
    write_to_dynamodb(rows)
    print(f"Wrote {len(rows)} records to S3 and DynamoDB")

File: initial_load.js
// Utility to export historical Google Ads click data from Snowflake using Node.js
// Requires the snowflake-sdk and aws-sdk packages.

const snowflake = require('snowflake-sdk');
const AWS = require('aws-sdk');

// --- Environment variables for Snowflake ---
const {
  SNOWFLAKE_ACCOUNT,
  SNOWFLAKE_USER,
  SNOWFLAKE_PASSWORD,
  SNOWFLAKE_WAREHOUSE,
  SNOWFLAKE_DATABASE,
  SNOWFLAKE_SCHEMA,
  S3_BUCKET,
  S3_KEY_PREFIX = 'click_performance/',
  DYNAMO_TABLE_NAME,
} = process.env;

const TABLE_NAME =
  'SEGMENT_EVENTS.GOOGLE_ADS_CLICK_PARAMETERS.CLICK_PERFORMANCE_REPORTS';

// --- AWS Clients ---
const s3 = new AWS.S3();
const ddb = new AWS.DynamoDB.DocumentClient();

function fetchAllRows() {
  return new Promise((resolve, reject) => {
    const connection = snowflake.createConnection({
      account: SNOWFLAKE_ACCOUNT,
      username: SNOWFLAKE_USER,
      password: SNOWFLAKE_PASSWORD,
      warehouse: SNOWFLAKE_WAREHOUSE,
      database: SNOWFLAKE_DATABASE,
      schema: SNOWFLAKE_SCHEMA,
    });

    connection.connect(err => {
      if (err) {
        reject(err);
        return;
      }
      connection.execute({
        sqlText: `SELECT * FROM ${TABLE_NAME}`,
        complete: (err, stmt, rows) => {
          connection.destroy();
          if (err) {
            reject(err);
          } else {
            resolve(rows);
          }
        },
      });
    });
  });
}

async function writeToDynamoDB(rows) {
  const batches = [];
  let current = [];
  for (const row of rows) {
    current.push({ PutRequest: { Item: row } });
    if (current.length === 25) {
      batches.push(current);
      current = [];
    }
  }
  if (current.length) batches.push(current);

  for (const batch of batches) {
    const params = {
      RequestItems: {
        [DYNAMO_TABLE_NAME]: batch,
      },
    };
    await ddb.batchWrite(params).promise();
  }
}

async function dumpToS3(rows, key) {
  const params = {
    Bucket: S3_BUCKET,
    Key: key,
    Body: JSON.stringify(rows),
  };
  await s3.putObject(params).promise();
}

async function main() {
  try {
    const rows = await fetchAllRows();
    const key = `${S3_KEY_PREFIX}initial_load.json`;
    await dumpToS3(rows, key);
    await writeToDynamoDB(rows);
    console.log(`Wrote ${rows.length} records to S3 and DynamoDB`);
  } catch (err) {
    console.error(err);
    process.exitCode = 1;
  }
}

if (require.main === module) {
  main();
}
